phabricatoremails/cli.py

```diff
@@ -0,0 +1,26 @@
+import argparse
+
+import sentry_sdk
+from phabricatoremails.config import Settings
+from phabricatoremails.init import init
+from phabricatoremails.service import service
+from statsd import StatsClient
+
+
+def cli():
+    settings = Settings.parse("settings.ini")
+    if settings.sentry_dsn:
+        sentry_sdk.init(settings.sentry_dsn)
+
+    parser = argparse.ArgumentParser()
+    commands = parser.add_subparsers(dest="command", required=True)
+
+    commands.add_parser("init", help="Initialize the database")
```

Comment:
I'm not sure that making `init` an optional part of regular startup is an improvement - for example, what if app initialization needs to happen in a different environment than real production, and we don't want to run the service just yet?

I think that making the `init` and `service` workflows mutually exclusive give us the most flexibility.

Reply:
> Ah fair point, you could turn them into two mutually exclusive groups and
still have them as booleans. My suggestion was mainly to remove the
redundancy of having free form text input and ambiguous arguments.

:thinking: Hmm, sorry, I don't quite understand.
Can you elaborate on how they the text is free form (it will error if the command isn't `init` or `service`), or how they're ambiguous arguments?

Reply:
To clarify, by "free form text" I meant that you are checking the value of an argument against string choices (i.e. you are comparing the value of `result.command` to a string, and this string is defined elsewhere -- in two places in this case.) This is a matter of preference I think, but I prefer to not duplicate strings in different places.

I think your approach is fine though I would prefer if the command strings were defined elsewhere.

And to improve self-documentation, perhaps adding a help text to `commands` to show the user what the positional argument is for would be good -- this would remove the ambiguity of the positional argument.

Thoughts?

Reply:
Ah, I see now what you mean by the string being defined twice, but I'm not sure how you'd avoid this. You could change it to `if args.init` instead of `if command == 'init'`, but you've still duplicated `"init"` from above. Additionally, now it's less obvious that the options are mutually exclusive, since you have two properties (`.init` and `.service`) that have their own on/off state.

-----

> And to improve self-documentation, perhaps adding a help text to commands to show the user what the positional argument is for would be good -- this would remove the ambiguity of the positional argument.

Ah, just to make sure we're on the same page, heads up that there's only one argument. It's called like this:
```
$ phabricator-emails init
# or
$ phabricator-emails service
```
If you forget the parameter, you get the following helpful error:
```
usage: phabricator-emails [-h] {init,service} ...
phabricator-emails: error: the following arguments are required: command
```

What kind of help text would you recommend to improve this?

Reply:
I believe if you use mutually exclusive groups you wouldn't have the problem you mentioned about mutual exclusivity.

Accessing an attribute on an object is quite different than comparing a variable to a string, and thus I wouldn't consider doing `args.init` duplication of `"init"`. 

It's not clear without the help text what the positional argument is for. I.e. "perform an action". 
```
usage: phabricator-emails [-h] {init,service}
positional arguments:
  {init,service}  <add help text here saying that these are the choices for an action to be performed>
...
```

You don't want the user to have to cause an error in order to see the instructions of a command (in this case, that the positional argument is in fact a command).

Reply:
I think there's two things we're discussing here: the help text, and free form text.

#### Help text

> You don't want the user to have to cause an error in order to see the instructions of a command (in this case, that the positional argument is in fact a command).

Ah, gotcha. If the user is confused by the available commands, then they can get more information using the standard procedure of `phabricator -h`:
```
$ phabricator-emails -h     
usage: phabricator-emails [-h] {init,service} ...

positional arguments:
  {init,service}
    init          Initialize the database
    service       Fetch events and send emails

optional arguments:
  -h, --help      show this help message and exit
```
#### free form text

> Accessing an attribute on an object is quite different than comparing a variable to a string, and thus I wouldn't consider doing args.init duplication of "init".

I'm not sure I agree here. When you do `parse_args()`, the return value isn't typed - it's essentially a dictionary. To me, all the following are equivalently "duplicating the string":
* `result.init` (when `result` is essentially a dictionary)
* `result['init']`
* `result == 'init'`

If it was being parsed to a more structured data type (in this case, an enum would be the best fit), then I'd see the benefit from moving away from a string.
But, you still have the parse the values in one way or another, so you're going to have `init`/`service` written twice.

What do you think?
Also, let's talk about this next week, stop working on your weekend :stuck_out_tongue:

Reply:
> Ah, gotcha. If the user is confused by the available commands, then they can get more information using the standard procedure of phabricator -h:

To clarify, I was referring to the lack of help text (or a title) to the positional arguments. I added a snippet below to illustrate (namely, I think using `title` here is sufficient). If the user does not cause an error, they don't see that the positional argument is in fact representing "commands".

> I'm not sure I agree here. When you do parse_args(), the return value isn't typed - it's essentially a dictionary. To me, all the following are equivalently "duplicating the string":

I would agree that the first two are very similar (whether accessing a key in a dict or a property/attribute of an object -- and both would be fine, although the way argparse works adds these as attributes and doesn't use a dict, so this is moot). Without splitting too many hairs, I would say that string comparison is quite different than accessing a property or key in a dict (as a simple example, both accessors will raise an exception if the key or attribute does not exist, there is no comparison in the third). 

Perhaps it's easier to illustrate via code what an alternate way would be that does not cause unnecessary repetition -- see below. I've added my earlier suggestion of separating `create_parser` for clarity as well.

```python
def create_parser():
    parser = argparse.ArgumentParser()

    # Add a title here to disambiguate what the positional arguments represent
    commands = parser.add_subparsers(
        dest="command", required=True, title="commands")

    # Define each parser individually so we can set default arguments
    init_parser = commands.add_parser("init", help="Initialize the database")
    service_parser = commands.add_parser(
        "service", help="Fetch events and send emails")

    # Set defaults (in this case we can reuse `func` and `args` without
    # replicating their values later on
    init_parser.set_defaults(func=init, params=(settings,))
    service_parser.set_defaults(func=service, params=(settings, stats))

    return parser


def cli():
    settings = Settings.parse("settings.ini")
    if settings.sentry_dsn:
        sentry_sdk.init(settings.sentry_dsn)
    stats = StatsClient()
    parser = create_parser()
    args = parser.parse_args()
    args.func(*args.params)
```

Enjoy your weekend!

Reply:
Oh, good call on `title="commands"`! What do you think about this?
```
$ phabricator-emails -h
usage: phabricator-emails [-h] {init,service} ...

optional arguments:
  -h, --help      show this help message and exit

commands:
  {init,service}
    init          Initialize the database
    service       Fetch events and send emails
```

-----

Ah, I now see the benefit you're talking about regarding the attributes vs string comparison - if you access the wrong property, you will be directly informed via exception that the property doesn't exist. So, taking this all together, and assuming we configure `argparse` correctly about mutually-exclusive arguments, we significantly decrease the likelihood of a silent programmer error in the command-handling section.

The part that's still rubbing me the wrong way is how we want the `argparse` logic to be separate from the command-handling logic (as recommended and :+1:'d earlier), yet this solution forces these to be _highly_ coupled.

What I mean is this:
```
args = parse_args()

...
### Command-handling

## What about if none/more than one are set?
# if not args.init and not args.service and not args.etc:
#     ...
# if {more than one are set}:
#     ...

if args.init:
    init()
elif args.service:
    service()
elif args.etc
    etc()
```

Obviously, we don't want/need that validation logic in the command-handling logic, since it's performed in `parse_args()`. However, the `args` data structure we receive doesn't make this constraint obvious - the usage of `args` is _coupled_ to `parse_args()`.

This implicit context makes this harder to maintain, since you don't know the shape of the `args` unless you read `parse_args()`.

I think I've got a good compromise here, though: what if we used a single property for the command, but we made sure that the value of that property is one of the values in our set of commands?

```
class Command(Enum):
    INIT = "init"
    SERVICE = "service"
    ETC = "etc"

def parse_command():
    ...
    args = parser.parse_args()
    return Command(args.command)

def cli():
    command = parse_command()
    ...
    if command == Command.INIT:
         init()
    elif command == Command.SERVICE:
        service()
    elif command == Command.ETC:
        etc()
    # Even without reading parse_command(), we know that there's only one
    # "command", and that it has to be set to one of the values of Command.
```
What do you think?

Reply:
> This implicit context makes this harder to maintain, since you don't know the shape of the args unless you read parse_args().

If I understand you correctly, this isn't necessarily an issue with the current needs of this module. There could be ways to handle some more complex cases with mutually exclusive groups, argument groups in general, additional subparsers, and more clever applications of `set_defaults` if needed. My example was to illustrate how we can factor out the functionality as it is currently present in this module.

We use [click](https://click.palletsprojects.com/en/7.x/) in other projects for managing `cli` (e.g. in [Lando API](https://github.com/mozilla-conduit/lando-api/blob/master/landoapi/cli.py)) which is also an option if these implementations become too complex.

Reply:
Ah, ok, stepping back a bit (I think I got caught up in the optional argument discussion from the beginning) what you're primarily advocating for is the moving of argument-handling logic to be together with with the argument-defining logic, is that correct?

So, rather than:
```
### define args
# <tell argparse that we accept an "init" command>
# <tell argparse that we accept a "service" command>
parser.parse_args()
...

### handle args
if <args command is init>:
    init()
elif <args command is service>:
    service()
...
# ! "init", "service" defined twice
```
You're looking for:
```
# <tell argparse that we accept an "init" command, in which case run "init()">
# <tell argparse that we accept an "service" command, in which case run "service()">
parser.parse_args()
# run associated command function
# ! "init", "service" defined once
```
Is that correct?

Reply:
Note that since we started this discussion, there's been an additional factor added: we need to parse the args before we can create `settings`, since `settings` depends on the `args.settings_file` argument.

So, if we try to define the parser commands and behaviour at the same time, it would look like:
```
def parse_command():
    parser = argparse.ArgumentParser()
    parser.add_argument("--settings-file", ...)

    commands = parser.add_subparsers(dest="command", required=True, ...)
    init_parser = commands.add_parser("init", help="Initialize the database")
    init_parser.set_defaults(func=lambda settings, _: init(settings))
    service_parser = commands.add_parser("service", help="Fetch events and send emails")
    service_parser.set_defaults(func=lambda settings, stats: service(settings, stats))
    return parser.parse_args()


def cli():
    args = parse_command()
    settings = Settings.parse(args.settings_file)
    # if settings.sentry_dsn: ...
    stats = StatsClient()
    args.func(settings, stats)
```
I feel that the `set_defaults` logic isn't super straightforward, since it's not especially clear how `func` is set. Additionally, you need to keep all the `func` lambdas in-sync (they need to accept every possible argument) which is a bummer.

Alternatively, we could handle this in a more straightforward procedural style:
```
def parse_command():
    parser = argparse.ArgumentParser()
    parser.add_argument("--settings-file", ...)

    commands = parser.add_subparsers(dest="command", required=True, ...)
    commands.add_parser("init", help="Initialize the database")
    commands.add_parser("service", help="Fetch events and send emails")
    return parser.parse_args()


def cli():
    args = parse_command()
    settings = Settings.parse(args.settings_file)
    # if settings.sentry_dsn: ...
    stats = StatsClient()
    if args.command == "init":
        init(settings)
    elif args.command == "service":
        service(settings, stats)
```
Though some `argparse` knowledge is required, the code path is more easily understandable.
Note that if it's the _string_ definitions that appear to be the code smell you're seeing, then we could throw them in `COMMAND_INIT`/`COMMAND_SERVICE` constants, though that relates to the discussion down [here](https://github.com/mozilla-conduit/phabricator-emails/pull/1#discussion_r404946084)

Reply:
An additional note here, you could factor out your arguments out of the parser entirely, which would still resolve the main issue of having repeated strings and boilerplate, but you should still be able to see how it logically makes sense to also factor out common arguments. Since your "commands" parser always maps to a method, it makes sense to factor it out as much as possible.

```python
def cli():
    parser = create_parser()
    args = parser.parse_args()

    # ...

    params = [settings]
    if args.func == service:
        params.append(stats)

    args.func(params)

```

Reply:
Ok, I see how you've factored out how all commands call a function, and how they all are called with `settings`.

I can visualize the difference between this and the current solution in-code:

_ | Existing code | Most recent recommendation
--|--|--
Parser section | `$command_string` | `$command_string => $function`
`cli()` section | `$command_string => $function($params)` | `$function => *$params`

I'm not sure that this axis of code reuse is improving maintainability, since it's making it tougher to realize the conditions in which each command function is called, and what parameters it'll receive.

To confirm that I'm on the same page, what problems is this solving?
* We no longer compare strings (instead, we're comparing function equility?)
* The function call and common parameters are re-used (though there's added `argparse` usage and param array mutation)

Reply:
Similar to our offline discussion, my suggested code is basically factoring out repeated things, and telling the user that `commands` is just calling a function with some parameters, as a pass-through, and nothing else is being done there. All three values do the exact same thing, which is calling a function with some arguments.